payu/telemetry.py

```diff
@@ -0,0 +1,276 @@
+
+        self.run_info = {}
+        self.run_info_filepath = None
+
+        # Check for config.yaml option to disable telemetry, and if an
+        # environment variable for an external telemetry config file is set
+        self.telemetry_enabled = (
+            config.get('telemetry', {}).get('enable', True)
+            and TELEMETRY_CONFIG in os.environ
+        )
+
+        self.scheduler = scheduler
+
+    def set_run_info_filepath(self, filepath: Path):
+        """This file path will be different depending depending if there
+        are model errors, whether archive runs or not. This is updated
+        in different stages the Experiment class"""
+        self.run_info_filepath = filepath
+
+    def set_run_info(self, experiment):
+        """Set the run information for the current run. This is
+        called in Experiment class after the model run is complete"""
+        self.run_info.update(get_metadata(experiment.metadata))
+        self.run_info.update(get_experiment_run_state(experiment))
+        self.run_info.update(get_manifests(experiment))
+
+    def clear_run_info(self):
+        self.run_info = {}
+
+    def record_run(self):
+        """
+        Build information for the current run and write it to a JSON file.
+        If telemetry is configured, post the telemetry job information
+        """
+        # Query the scheduler just before recording the run information to
+        # try get the most up-to-date information of the usage statistics
+        # as they only get updated periodically
+        self.run_info.update(get_scheduler_run_info(self.scheduler))
+
+        # Write run job information to a JSON file
+        if self.run_info_filepath is None:
+            warnings.warn(
+                "Run job output file is not defined"
+            )
+        else:
+            with open(self.run_info_filepath, 'w', encoding='utf-8') as f:
+                json.dump(self.run_info, f, ensure_ascii=False, indent=4)
+
+        if not self.telemetry_enabled:
+            # Skip any external telemetry
+            return
+
+        # Check for valid external telemetry configuration file
+        external_config = get_external_telemetry_config()
+        if external_config is None:
+            # Skip any external telemetry
+            return
+
+        # Add hostname to the run info fields
+        self.run_info.update({'hostname': external_config[HOSTNAME_FIELD]})
+
+        starttime = datetime.now()
+
+        # Using threading to run the one post request in the background
+        thread = threading.Thread(
+            target=post_telemetry_data,
+            kwargs= {
+                'url': external_config[TELEMETRY_URL_FIELD],
+                'token': external_config[TELEMETRY_TOKEN_FIELD],
+                'data': self.run_info,
+                'service_name': external_config[TELEMETRY_SERVICE_NAME_FIELD],
+                'host': external_config[TELEMETRY_HOST_FIELD],
+            },
+        )
```

Comment:
```suggestion
            kwargs= {
                'url': external_config[TELEMETRY_CONFIG_FIELD['URL']],
                'token': external_config[TELEMETRY_CONFIG_FIELD['TOKEN']],
                'data': self.run_info,
                'service_name': external_config[TELEMETRY_CONFIG_FIELD['SERVICE_NAME']],
                'host': external_config[TELEMETRY_CONFIG_FIELD['TELEMETRY_HOST']],
            },
        )
```
These would need to be modified if changed to a `dict`. This looks a little more unwieldy than the previous approach ...